File: app/src/main/java/de/jerleo/database/TableBill.java
package de.jerleo.database;

import android.content.ContentValues;
import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

import de.jerleo.android.DateHelper;
import de.jerleo.database.DatabaseColumn.DataType;
import de.jerleo.model.Bill;

import static android.provider.BaseColumns._ID;
import static de.jerleo.database.Constants.BEGIN;
import static de.jerleo.database.Constants.BILL;
import static de.jerleo.database.Constants.DESCRIPTION;

public class TableBill extends DatabaseTable {

    public TableBill(Database database) {

        super(database, BILL, true);

        addColumn(new DatabaseColumn(DESCRIPTION, DataType.TEXT));
        addColumn(new DatabaseColumn(BEGIN, DataType.TEXT));
    }

    @Override
    public ContentValues getValues(Object object, boolean forUpdate) {

        final Bill header = (Bill) object;

        final ContentValues values = new ContentValues();
        if (forUpdate)
            values.put(_ID, header.getId());
        values.put(DESCRIPTION, header.getDescription());
        values.put(BEGIN, DateHelper.getDatabaseDate(header.getBegin()));

        return values;
    }

    @Override
    public String getWhereClause(Object object) {

        final Bill header = (Bill) object;
        return _ID + " = " + header.getId();
    }

    @Override
    public List<Bill> read(String where) {

        final ArrayList<Bill> bills = new ArrayList<>();
        final String[] columns = {_ID, DESCRIPTION, BEGIN};
        final Cursor cursor = database.query(BILL, columns, where, null, null,
                null, null);
        while (cursor.moveToNext()) {
            final Bill bill = new Bill();
            int col = 0;
            bill.setId(cursor.getLong(col++));
            bill.setDescription(cursor.getString(col++));
            bill.setBegin(DateHelper.getDatabaseDate(cursor.getString(col)));
            bills.add(bill);
        }
        return bills;
    }

}

File: app/src/main/kotlin/de/jerleo/database/table/TableBill.kt
package de.jerleo.database.table

import android.content.ContentValues
import android.provider.BaseColumns
import de.jerleo.android.DateHelper
import de.jerleo.database.Column
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Table
import de.jerleo.model.Bill

class TableBill(database: Database) :
    Table(database, Constants.BILL, true) {

    init {
        add(Column(Constants.DESCRIPTION, Column.Type.TEXT))
        add(Column(Constants.BEGIN, Column.Type.TEXT))
    }

    override fun values(obj: Any, forUpdate: Boolean): ContentValues {
        val header = obj as Bill
        val values = ContentValues()
        if (forUpdate) values.put(BaseColumns._ID, header.id)
        values.put(Constants.DESCRIPTION, header.name)
        values.put(Constants.BEGIN, DateHelper.format(header.dateFrom))
        return values
    }

    override fun whereClause(obj: Any): String = BaseColumns._ID + " = " + (obj as Bill).id

    override fun read(condition: String?): ArrayList<Bill> {
        val bills = ArrayList<Bill>()
        val columns = arrayOf(BaseColumns._ID, Constants.DESCRIPTION, Constants.BEGIN)
        val cursor = database.query(
            Constants.BILL, columns, condition, null, null,
            null, null
        )
        while (cursor.moveToNext()) {
            val bill = Bill()
            var col = 0
            bill.id = cursor.getLong(col++)
            bill.name = cursor.getString(col++)
            bill.dateFrom = DateHelper.parse(cursor.getString(col))
            bills.add(bill)
        }
        return bills
    }

    fun import(): MutableList<Bill> {
        val bills: MutableList<Bill> = read(null)
        for (bill in bills) {
            val id = bill.id.toString()
            val items = Database.tableItem.read(id)
            bill.items = items
        }
        return bills
    }
}